### PR TITLE
250 fsm state diagram mermaid generator

### LINKED ISSUE #250: FSM State diagram generator in mermaid
### Motivation

It's easier if we can visualize the state diagram in the FSM from the implementation. 

### Desired solution

A mermaid intermediary that represent FSM state diagram.

### Alternatives considered

_No response_

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by quekyj**
## Description & Motivation

It's easier if we can visualize the state diagram in the FSM from the implementation.

## Related Issue(s)

Fix #250 

## Testing

Added unit test to make sure the diagram is generated and there the word "mermaid" inside the file.

## Backwards-compatibility

Yes. 

## Documentation

Yes. Added the documentation.

### CODE REVIEW

File: example/counter.dart [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/// Copyright (C) 2021 Intel Corporation
```
Comment (mkorbel1) [inline example/counter.dart:1]: Why is this file here?  This looks like a copy of example.dart?

File: example/oven_fsm.dart [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    // add time elapsed Counter
+    final counterReset = Logic(name: 'counter_reset');
+    final en = Logic(name: 'counter_en');
+    final counter = Counter(en, counterReset, clk, name: 'counter_module');
+
+    states = [
+      // Standby State
+      State<OvenStates>(OvenStates.standby, events: {
+        button.eq(Button.start()): OvenStates.cooking,
```
Comment (mkorbel1) [inline example/oven_fsm.dart:46]: Let's try naming the `Logic`s that we pass into there so that the FSM diagram has some more interesting stuff?
```dart
Logic(name: 'button_start')..gets(button.eq(Button.start()))
```

File: example/oven_fsm.dart [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        button.eq(Button.start()): OvenStates.cooking
+      }, actions: [
+        led < LEDLight.green().value,
+        counterReset < 1,
+        en < 0,
+      ])
+    ];
+
+    StateMachine<OvenStates>(clk, reset, OvenStates.standby, states)
+        .generateDiagram(outputPath: 'oven_fsm.md');
```
Comment (mkorbel1) [inline example/oven_fsm.dart:83]: We should call the `generateDiagram` somehow from within `main` maybe?

File: example/oven_fsm.dart [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  WaveDumper(oven, outputPath: 'example/oven.vcd');
+
+  Simulator.registerAction(120, () {
+    // ignore: avoid_print
+    print('Simulation End');
+  });
+
+  Simulator.setMaxSimTime(120);
+
```
Comment (mkorbel1) [inline example/oven_fsm.dart:124]: We should include some `print` messages (like in example.dart) explaining that we can look at the waves, how to look at the mermaid diagram, etc.

File: lib/src/state_machine.dart [hunk 5]
```diff
@@ -117,3 +135,53 @@ class State<StateIdentifier> {
   /// and [actions] associated with that state.
   State(this.identifier, {required this.events, required this.actions});
 }
+
+/// A state diagram generator for FSM.
+///
+/// Outputs to vcd format at [outputPath].
+class MermaidStateDiagram {
```
Comment (mkorbel1) [inline lib/src/state_machine.dart:142]: Should this be a private class or do you intend to expose this as a public api?
`_MermaidStateDiagram`

File: lib/src/state_machine.dart [hunk 6]
```diff
@@ -117,3 +135,53 @@ class State<StateIdentifier> {
...
+    diagram = 'stateDiagram\n';
+  }
+
+  /// Register a new state to the mermaid diagram object.
+  void addState(String state) => diagram = '$diagram\n$state\n';
+
+  /// Register a new transition [event] that point the
+  /// current state [currState] to next state [nextState].
+  void addTransitions(String currState, String nextState, String event) =>
+      diagram = '$diagram$currState --> $nextState: $event\n';
```
Comment (mkorbel1) [inline lib/src/state_machine.dart:167]: I think we can use a `StringBuffer` instead of re-setting the `diagram` variable each time

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: example/counter.dart (hunk 1 above)
Comment (quekyj) [reply example/counter.dart:1]: Yes, let me change the dependent to `'import ./example.dart'` instead.

File: example/oven_fsm.dart (hunk 2 above)
Comment (quekyj) [reply example/oven_fsm.dart]: Yup, sure!

File: example/oven_fsm.dart (hunk 3 above)
Comment (quekyj) [reply example/oven_fsm.dart:84]: Ya, let me create a getter instead.

File: lib/src/state_machine.dart (hunk 5 above)
Comment (quekyj) [reply lib/src/state_machine.dart:142]: Ya, this should be private class instead.

File: example/oven_fsm.dart [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    // Module logic must consume registered inputs and output to registered
+    // outputs. `led` output also added as the output port.
+    button = addInput('button', button, width: button.width);
+    reset = addInput('reset', reset);
+    final led = addOutput('led', width: button.width);
+
+    // An internal clock generator
+    final clk = SimpleClockGenerator(10).clk;
+
+    // Register inputs and outputs, `counterReset` and `en` for internal signals
```
Comment (chykon) [inline example/oven_fsm.dart:64]: If this is the registration of internal signals, then, I think, the text `Register inputs and outputs` is not needed? As far as I understand, registration is `addInput`/`addOutput`, but here it looks like `Define local signals` from <https://github.com/intel/rohd/blob/58584d2bc6ac0d7844dd92b2ecc85a5365bcfa5c/example/example.dart#L60-L62> I can be wrong.

File: example/oven_fsm.dart [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    // A list of `OvenStates` that describe the FSM. Note that
+    // `OvenStates` consists of identifier, events and actions. We
+    // can think of `identifier` as the state name, `events` is a map of event
+    // that trigger next state. `actions` is the behaviour of current state,
+    // like what is the actions need to be shown separate current state with
+    // other state. Represented as List of conditionals to be executed.
+    final states = [
+      // identifier: standby state, represent by `OvenStates.standby`.
+      State<OvenStates>(OvenStates.standby,
+          // events: When the button `start` is pressed during standby state,
```
Comment (chykon) [inline example/oven_fsm.dart:82]: Maybe move `events:` separately up to a new line, as done in the code below?

File: example/oven_fsm.dart [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+          // OvenState will changed to `OvenStates.cooking` state.
+          events: {
+            Logic(name: 'button_start')..gets(button.eq(Button.start())):
+                OvenStates.cooking,
+          },
+          // actions: During the standby state, `led` is change to blue; timer's
```
Comment (chykon) [inline example/oven_fsm.dart:88]: Maybe move `actions:` separately up to a new line, as done in the code below?

File: example/oven_fsm.dart [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+    // Assign the oven StateMachine object to public variable declared.
+    oven = StateMachine<OvenStates>(clk, reset, OvenStates.standby, states);
+  }
+
+  // An ovenStateMachine that represent in getter.
+  StateMachine<OvenStates> get ovenStateMachine => oven;
+}
+
+void main() async {
```
Comment (chykon) [inline example/oven_fsm.dart:168]: Maybe use `Future<void> main ...` like in the other examples?

File: example/oven_fsm.dart [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+  // to build it.
+  await oven.build();
+
+  // Now let's try simulating!
+
+  // Let's start off with asserting reset to Oven.
+  reset.inject(1);
+
+  // Attach a waveform dumper so we can see what happens.
+  WaveDumper(oven, outputPath: 'example/oven.vcd');
```
Comment (chykon) [inline example/oven_fsm.dart:196]: Perhaps it is better to output to the root of the project, where `oven_fsm.md` will already be.

File: example/oven_fsm.dart [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+  // Drop reset at time 25.
+  Simulator.registerAction(25, () => reset.put(0));
+
+  // Press button `00` => start at time 25.
+  Simulator.registerAction(25, () {
+    button.put(bin('00'));
```
Comment (chykon) [inline example/oven_fsm.dart:203]: Can use `Button.start()` instead of `bin('00')`? Might need to update the comments.

File: example/oven_fsm.dart [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+  });
+
+  // Press button `01` => pause at time 50.
+  Simulator.registerAction(50, () {
+    button.put(bin('01'));
```
Comment (chykon) [inline example/oven_fsm.dart:208]: Can use `Button.pause()` instead of `bin('01')`? Might need to update the comments.

File: example/oven_fsm.dart [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+  });
+
+  // Press button `10` => resume at time 70.
+  Simulator.registerAction(70, () {
+    button.put(bin('10'));
```
Comment (chykon) [inline example/oven_fsm.dart:213]: Can use `Button.resume()` instead of `bin('10')`? Might need to update the comments.

File: example/oven_fsm.dart [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+  });
+
+  // Print a message when we're done with the simulation!
+  Simulator.registerAction(120, () {
+    print('Simulation End');
```
Comment (chykon) [inline example/oven_fsm.dart:218]: If the <https://github.com/intel/rohd/pull/267> changes are accepted, then you can replace `print` with `stdout.writeln`.

**Review by chykon — COMMENTED** [review]
Great! I really liked the example. Here are my suggestions:

File: oven_fsm.md [hunk 16]
```diff
@@ -0,0 +1,11 @@
+```mermaid
```
[the remark this answers is not in the document]
Comment (quekyj) [reply oven_fsm.md]: Ya, you are right. Let me remove from git

File: oven_fsm.md [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+stateDiagram
+
+[*] --> OvenStates.standby
+OvenStates.standby --> OvenStates.cooking: button_start
+OvenStates.cooking --> OvenStates.paused: button_pause
+OvenStates.cooking --> OvenStates.completed: counter_time_complete
+OvenStates.paused --> OvenStates.cooking: button_resume
+OvenStates.completed --> OvenStates.cooking: button_start
+
```
[the remark this answers is not in the document]
Comment (quekyj) [reply oven_fsm.md]: Nope, let me remove this in the implementation.

File: oven_fsm.md [hunk 18]
```diff
@@ -0,0 +1,11 @@
+```mermaid
+stateDiagram
```
[the remark this answers is not in the document]
Comment (quekyj) [reply oven_fsm.md]: Let me change to -v2. I didn't realize there is a version 2 for this. 😄

File: example/oven_fsm.dart [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+  LEDLight.yellow() : this._(bin('00'));
+  LEDLight.blue() : this._(bin('01'));
+  LEDLight.red() : this._(bin('10'));
+  LEDLight.green() : this._(bin('11'));
+}
+
+// Define a class OvenModule that extends ROHD's abstract Module class
+class OvenModule extends Module {
+  // A public variable with type StateMachine<OvenStates> `oven`.
+  // We want to return this variable to the main module for flexbility.
```
[the remark this answers is not in the document]
Comment (quekyj) [reply example/oven_fsm.dart]: Ya. 🤣

File: example/oven_fsm.dart [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+/// SPDX-License-Identifier: BSD-3-Clause
+///
+/// oven_fsm.dart
+/// A simple oven FSM implementation using ROHD.
+///
+/// 2023 February 13
+/// Author: Yao Jing Quek <yao.jing.quek@intel.com>
+///
+
+// ignore_for_file: avoid_print
```
[the remark this answers is not in the document]
Comment (quekyj) [reply example/oven_fsm.dart:11]: From the comment in #267. Looks like we want to stick with `print` function instead to motivate the linting check.

File: example/oven_fsm.dart [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+
+// Import the ROHD package
+import 'package:rohd/rohd.dart';
+
+// Import the counter module implement in example.dart
+import './example.dart';
+
+// Enumerated type named `OvenStates` with four possible states:
+// `standby`, `cooking`,`paused`, and `completed`.
+enum OvenStates { standby, cooking, paused, completed }
```
[the remark this answers is not in the document]
Comment (quekyj) [reply example/oven_fsm.dart]: Ya, you are right. Interesting, I just know the fact that `enum` need to be singular.

File: example/oven_fsm.dart (hunk 7 above)
Comment (quekyj) [reply example/oven_fsm.dart]: Now you mentioned it, it seems like a local signal to me. Let me change the wording.

File: example/oven_fsm.dart (hunk 10 above)
Comment (quekyj) [reply example/oven_fsm.dart]: Ya, can do that.

File: example/oven_fsm.dart (hunk 15 above)
Comment (quekyj) [reply example/oven_fsm.dart:220]: From the comment in https://github.com/intel/rohd/pull/267. Looks like we want to stick with print function instead to motivate the linting check.

**Comment by quekyj** [comment]
@mkorbel1 @chykon Let me know if still need any changes! :slightly_smiling_face:

File: test/fsm_test.dart [hunk 22]
```diff
@@ -149,6 +162,20 @@ void main() {
...
+
+void verifyMermaidStateDiagram(String filePath) {
+  // check if the diagram exist
+  final file = File(filePath);
+  final existDiagram = file.existsSync();
+  expect(existDiagram, isTrue);
+
+  // check if the file generated is mermaid file
+  final fileContents = file.readAsStringSync();
+  expect(fileContents, contains('mermaid'));
```
Comment (chykon) [inline test/fsm_test.dart:178]: How about tightening up the check: ` ```mermaid\nstateDiagram-v2\n    ` (with four spaces at the end as an indent)?

**Review by chykon — COMMENTED** [review]
Great job! Here's something else:

File: example/oven_fsm.dart [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+
+  // Build an Oven Module and passed the `button` and `reset`.
+  final oven = OvenModule(button, reset);
+
+  // Generate a Mermaid FSM diagram and save as the name `oven_fsm.md`.
+  // Note that the extension of the files is recommend as .md or .mmd.
+  //
+  // Check on https://mermaid.js.org/intro/ to view the diagram generated.
+  // If you are using vscode, you can download the mermaid extension.
+  oven.ovenStateMachine.generateDiagram(outputPath: 'oven_fsm.md');
```
[the remark this answers is not in the document]
Comment (quekyj) [reply example/oven_fsm.dart:186]: ya, can do that.

File: lib/src/state_machine.dart [hunk 24]
```diff
@@ -117,3 +137,52 @@ class State<StateIdentifier> {
...
+  /// mermaid as `stateDiagram`.
+  ///
+  /// Passed output path to save in custom directory.
+  _MermaidStateDiagram({this.outputPath = 'stateDiagram.md'})
+      : _outputFile = File(outputPath) {
+    diagram = StringBuffer('stateDiagram-v2');
+  }
+
+  /// Register a new state to the mermaid diagram object.
+  void addState(String state) => diagram.write('\n$state\n');
```
[the remark this answers is not in the document]
Comment (quekyj) [reply lib/src/state_machine.dart:168]: Hmm, I think better remove this since the function `addTransitions()` already add state.

File: lib/src/state_machine.dart [hunk 25]
```diff
@@ -117,3 +137,52 @@ class State<StateIdentifier> {
   /// and [actions] associated with that state.
   State(this.identifier, {required this.events, required this.actions});
 }
+
+/// A state diagram generator for FSM.
+///
+/// Outputs to vcd format at [outputPath].
+class _MermaidStateDiagram {
+  /// The diagram to be return as String
+  late StringBuffer diagram;
```
[the remark this answers is not in the document]
Comment (quekyj) [reply lib/src/state_machine.dart:147]: Ya, I can make it private.

File: lib/src/state_machine.dart [hunk 26]
```diff
@@ -99,6 +101,24 @@ class StateMachine<StateIdentifier> {
       )
     ]);
   }
+
+  /// Generate a FSM state diagram [_MermaidStateDiagram].
```
[the remark this answers is not in the document]
Comment (quekyj) [reply lib/src/state_machine.dart:106]: Ya, I think we can do that. Because even though it is a private element, people still can read the function for it right?

From effective dart docs:
> CONSIDER writing doc comments for private APIs.
Doc comments aren’t just for external consumers of your library’s public API. They can also be helpful for understanding private members that are called from other parts of the library.

File: test/fsm_test.dart [hunk 27]
```diff
@@ -33,7 +35,9 @@ class TestModule extends Module {
         b < ~c,
       ]),
     ];
-    StateMachine<MyStates>(clk, reset, MyStates.state1, states);
+
+    StateMachine<MyStates>(clk, reset, MyStates.state1, states)
+        .generateDiagram(outputPath: 'simple_fsm.md');
```
[the remark this answers is not in the document]
Comment (quekyj) [reply test/fsm_test.dart]: Yup, thanks for pointing that out.

**Comment by quekyj** [comment]
@chykon the code looks a lot cleaner now 🤣.

File: lib/src/state_machine.dart (hunk 26 above)
[the remark this answers is not in the document]
Comment (chykon) [reply lib/src/state_machine.dart:106]: Yes you are right. Let's take care of the quality of documentation for developers too.

**Review by chykon — APPROVED** [review]
Great job! I'm just learning how to do code reviews, so I'm sorry if I was not quite correct somewhere or pointed out very minor details. I hope that helped you look from the outside 😄

File: example/oven_fsm.dart [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+import './example.dart';
+
+// Enumerated type named `OvenState` with four possible states:
+// `standby`, `cooking`,`paused`, and `completed`.
+enum OvenState { standby, cooking, paused, completed }
+
+// One-hot encoded `Button` that extends from Const() value.
+// Represent start, pause, and resume as constant value `00`, `01`,
+// and `10` respectively.
+class Button extends Const {
```
Comment (mkorbel1) [inline example/oven_fsm.dart:26]: I think we can replace both `Button` and `LEDLight` with `enum`s with values:
https://dart.dev/guides/language/language-tour#enumerated-types

I think it would look cleaner than having these `Const` subclasses

File: example/oven_fsm.dart [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+  LEDLight.green() : this._(bin('11'));
+}
+
+// Define a class OvenModule that extends ROHD's abstract Module class.
+class OvenModule extends Module {
+  // A public variable with type StateMachine<OvenState> `oven`.
+  // We want to return this variable to the main module for flexibility.
+  // Use `late` to indicate that the value will not be null
+  // and will be assign in the later section.
+  late StateMachine<OvenState> oven;
```
Comment (mkorbel1) [inline example/oven_fsm.dart:50]: This could be private since you have a public getter already.  It would be nice to put the getter next to this variable definition

File: example/oven_fsm.dart [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+  await oven.build();
+
+  // Now let's try simulating!
+
+  // Let's start off with asserting reset to Oven.
+  reset.inject(1);
+
+  // Attach a waveform dumper so we can see what happens.
+  WaveDumper(oven, outputPath: 'oven.vcd');
+
```
Comment (mkorbel1) [inline example/oven_fsm.dart:199]: It could be fun to register something listening to the LED (`led.changed.listen...`) so that without looking at the waves you can still see the oven going through different states.

File: example/oven_fsm.dart [hunk 31]
```diff
@@ -0,0 +1,228 @@
+/// Copyright (C) 2023 Intel Corporation
```
Comment (chykon) [inline example/oven_fsm.dart:1]: I think it will also be necessary to add `oven_fsm.dart` to `test/example_test.dart`.

File: example/oven_fsm.dart [hunk 32]
```diff
@@ -0,0 +1,228 @@
...
+
+    // Assign the oven StateMachine object to public variable declared.
+    oven = StateMachine<OvenState>(clk, reset, OvenState.standby, states);
+  }
+
+  // An ovenStateMachine that represent in getter.
+  StateMachine<OvenState> get ovenStateMachine => oven;
+}
+
+Future<void> main() async {
```
Comment (chykon) [inline example/oven_fsm.dart:170]: If this example is added to `test/example_test.dart`, then the `noPrint` parameter must be added like the other examples.

File: example/oven_fsm.dart [hunk 33]
```diff
@@ -0,0 +1,228 @@
...
+  });
+
+  // Press button resume => `10` at time 70.
+  Simulator.registerAction(70, () {
+    button.put(Button.resume().value);
+  });
+
+  // Print a message when we're done with the simulation!
+  Simulator.registerAction(120, () {
+    print('Simulation End');
```
Comment (chykon) [inline example/oven_fsm.dart:220]: If this example is added to `test/example_test.dart`, then `print` must be wrapped in `if (!noPrint)` like the other examples.

**Review by chykon — COMMENTED** [review]
Maybe add `oven_fsm.dart` to `test/example_test.dart`?

**Review by chykon — COMMENTED** [review]
Tiny notes:

File: example/oven_fsm.dart [hunk 34]
```diff
@@ -20,34 +20,40 @@ import './example.dart';
...
+  const LEDLight({required this.value});
+
+  final int value;
 }
 
 // Define a class OvenModule that extends ROHD's abstract Module class.
 class OvenModule extends Module {
-  // A public variable with type StateMachine<OvenState> `oven`.
-  // We want to return this variable to the main module for flexibility.
+  // A private variable with type StateMachine<OvenState> `oven`.
```
[the remark this answers is not in the document]
Comment (quekyj) [reply example/oven_fsm.dart:52]: Ya. Should change that 🤣

File: example/oven_fsm.dart [hunk 35]
```diff
@@ -0,0 +1,252 @@
...
+          // `counterReset` is set to 1 (Reset value);
+          // timer's `en` is set to 0 (Disable value update).
+          actions: [
+            led < LEDLight.green.value,
+            counterReset < 1,
+            en < 0,
+          ])
+    ];
+
+    // Assign the _oven StateMachine object to public variable declared.
```
Comment (chykon) [inline example/oven_fsm.dart:176]: Sorry, I didn't explain correctly. I wanted to point out that perhaps the `public variable declared` part should be updated as the variable is now private. Are there any problems in the `oven StateMachine` part?

Comment (quekyj) [reply example/oven_fsm.dart:179]: I see what you mean now. Should be changed to private variable declared instead.

File: .gitignore [hunk 36]
```diff
@@ -12,6 +12,8 @@ doc/api/
 # Other stuff
 tmp_test/
 *.vcd
+.vscode/
```
Comment (chykon) [inline .gitignore:15]: Extra `.vscode/` in .gitignore rules.

**Review by mkorbel1 — APPROVED** [review]
This looks great, awesome contribution, thank you!
